driver/labjack/writer_sink.cpp

```diff
@@ -0,0 +1,277 @@
+    state_map(state_map),
+    state_index_keys(state_index_keys) {
+    this->timer = loop::Timer(this->state_rate);
+}
+
+
+std::pair<synnax::Frame, freighter::Error> labjack::StateSource::read(
+    breaker::Breaker &breaker) {
+    std::unique_lock<std::mutex> lock(this->state_mutex);
+    this->timer.wait(breaker); // sleep for state period
+    waiting_reader.wait_for(lock, this->state_rate.period().chrono());
+    return std::make_pair(this->get_state(), freighter::NIL);
+}
+
+synnax::Frame labjack::StateSource::get_state() {
+    // frame size = # monitored states + # index channels for the states
+    auto state_frame = synnax::Frame(this->state_map.size() + this->state_index_keys.size());
+
+    for (auto key: this->state_index_keys) {
+        state_frame.add(
+            key,
+            synnax::Series(
+                synnax::TimeStamp::now().value,
+                synnax::TIMESTAMP
+            )
+        );
+    }
+    for (auto &[key, value]: this->state_map) {
+        auto s = val_to_series(value.state, value.data_type);
+        state_frame.add(
+            value.state_key,
+            std::move(s)
+        );
+    }
+
+    return state_frame;
+}
+
+void labjack::StateSource::update_state(synnax::Frame frame) {
+    std::unique_lock<std::mutex> lock(this->state_mutex);
+    auto frame_index = 0;
+    for (auto key: *(frame.channels)) {
+        if (std::find(
+                state_index_keys.begin(),
+                state_index_keys.end(),
+                key
+            ) != state_index_keys.end())
+            continue;
+
+        double value = series_to_val(frame.series->at(frame_index));
+        this->state_map[key].state = value;
+        frame_index++;
+    }
+
+    waiting_reader.notify_one();
+}
+
+///////////////////////////////////////////////////////////////////////////////////
+//                                   WriteSink                                   //
+///////////////////////////////////////////////////////////////////////////////////
+labjack::WriteSink::WriteSink(
+    const std::shared_ptr<task::Context> &ctx,
+    const synnax::Task &task,
+    const labjack::WriterConfig &writer_config
+) : ctx(ctx),
+    task(task),
+    writer_config(writer_config) {
+    auto breaker_config = breaker::Config{
+        .name = task.name,
+        .base_interval = 1 * SECOND,
+        .max_retries = 20,
+        .scale = 1.2,
+    };
+
+    this->breaker = breaker::Breaker(breaker_config);
+
+    auto state_index_keys = this->get_index_keys(); // retrieve state index from first state channel
+
+    this->state_source = std::make_shared<labjack::StateSource>(
+        this->writer_config.state_rate,
+        state_index_keys,
+        this->writer_config.initial_state_map
+    );
+}
+
+labjack::WriteSink::~WriteSink() {
+    this->stop("");
+}
+
+void labjack::WriteSink::init() {
+    if (this->writer_config.device_type == "") {
+        auto [dev, err] = this->ctx->client->hardware.retrieveDevice(
```

Comment:
Is this the only place you retrieve the device? If so, shouldn't the failure to retrieve the device be communicated via task state?

Reply:
Also `retrieveDevice` should be snake case